tests/saw/spec/handshake/handshake_io_lowlevel.saw

```diff
@@ -98,22 +97,33 @@ let conn_handshake_message_number pconn =
 let conn_chain_and_key pconn =
     crucible_field (crucible_field pconn "handshake_params") "our_chain_and_key";
 
+//conn->psk_params.chosen_psk
+let conn_chosen_psk pconn =
+    crucible_field (crucible_field pconn "psk_params") "chosen_psk";
+
+//conn->early_data_state
+let conn_early_data_state pconn = crucible_field pconn "early_data_state";
+
 // Ghost state that represents the number of times the connection write socket
 // has been corked/uncorked. 
 corked <- crucible_declare_ghost_state "corked";
 crucible_ghost_value corked {{ 0 : [2] }};
 
 // setup_handshake de-serializes parts of the s2n_handshake and s2n_connection
 // structs into a Cryptol record. It also deserializes the ghost state.
-let setup_connection = do {
+//
+// The chosen_psk_null argument is true if conn->psk_params.chosen_psk should
+// be set to NULL, which enables the use of FULL_HANDSHAKE. If false, then
+// chosen_psk is set to a non-NULL value, disabling FULL_HANDSHAKE.
+let setup_connection chosen_psk_null = do {
```

Comment:
Why can't the proof work for both cases? Why do we need to choose one or the other to prove?

Reply:
Could you always allocate chosen_psk, but branch on a boolean value indicating whether it's null for the proof or not? So it's never technically null in the proofs, but we sometimes ignore it.

Reply:
The one place where this would pose an issue is in `s2n_conn_set_handshake_type_spec`, which actually simulates the C code corresponding to `s2n_conn_set_handshake_type`. If we always allocate `chosen_psk` in the SAW specification, then SAW would never check any of the TLS 1.3 code paths where `chosen_psk` is `NULL` (i.e., where the handshake type enables `FULL_HANDSHAKE`). This seemed undesirable to me, which is why I went the current approach. I could change this, I suppose, but then `s2n_conn_set_handshake_type_spec` really ought to come with a giant disclaimer about the fact that it wouldn't simulate certain code paths.

Reply:
I've opened #3052 as a reminder to look into removing the need for this `chosen_psk_null` argument once SAW fixes GaloisInc/saw-script#1161.